ddev/src/ddev/utils/scripts/check_pr.py

```diff
@@ -0,0 +1,189 @@
+                    added[i - removed] = line[1:]
+                elif line.startswith('-'):
+                    removed += 1
+
+    return files
+
+
+def get_missing_changelogs(git_diff: str, suffix: str) -> None:
+    targets = {}
+    for filename, lines in get_added_lines(git_diff).items():
+        target, _, path = filename.partition('/')
+        if not path:
+            continue
+
+        targets.setdefault(target, {})[path] = lines
+
+    errors: tuple[str, int, str] = []
+    for target, files in sorted(targets.items()):
+        if not requires_changelog(target, files.keys()):
+            continue
+
+        changelog_file = 'CHANGELOG.md'
+        if changelog_file not in files:
+            errors.append((f'{target}/{changelog_file}', 1, 'Missing changelog entry'))
+            continue
+
+        added_lines = files[changelog_file]
+        line_numbers_missing_suffix = []
+        lines_with_suffix = 0
+        for line_number, line in added_lines.items():
+            if not line.startswith('* '):
+                continue
+            elif line.endswith(suffix):
+                lines_with_suffix += 1
+            else:
+                line_numbers_missing_suffix.append(line_number)
+
+        if lines_with_suffix == len(line_numbers_missing_suffix) == 0:
+            errors.append((f'{target}/{changelog_file}', 1, 'Missing changelog entry'))
+        elif line_numbers_missing_suffix:
+            for line_number in line_numbers_missing_suffix:
+                errors.append(
+                    (
+                        f'{target}/{changelog_file}',
+                        line_number,
+                        f'The first line of every new changelog entry must '
+                        f'end with the associated PR number `{suffix}`',
+                    )
+                )
+
+    return errors
+
+
+def changelog_impl(*, ref: str, diff_file: str, pr_number: int) -> None:
+    on_ci = os.environ.get('GITHUB_ACTIONS') == 'true'
+    if on_ci:
+        with open(diff_file, encoding='utf-8') as f:
+            git_diff = f.read()
+    else:
+        git_diff = git('diff', f'{ref}...')
+
+    errors = get_missing_changelogs(git_diff, f' (#{pr_number})')
+    if not errors:
+        return
+    elif os.environ.get('GITHUB_ACTIONS') == 'true':
+        for relative_path, line_number, message in errors:
+            message = '%0A'.join(message.splitlines())
+            print(f'::error file={relative_path},line={line_number}::{message}')
+    else:
+        for relative_path, line_number, message in errors:
+            print(f'{relative_path}, line {line_number}: {message}')
+
+    sys.exit(1)
+
+
+def changelog_command(subparsers) -> None:
+    parser = subparsers.add_parser('changelog')
+    parser.add_argument('--ref', default='origin/master')
+    parser.add_argument('--diff-file')
+    parser.add_argument('--pr-number', type=int, default=1)
+    parser.set_defaults(func=changelog_impl)
```

Comment:
```suggestion
    # We associate a command function with the subcommand parser. This allows us to emulate click's command group behavior.
    parser.set_defaults(func=changelog_impl)
```
This is neat, but I needed a minute to understand its purpose which involved also jumping to the `main` function. Overengineering concerns aside, having the intent stated right here would improve readability imo.